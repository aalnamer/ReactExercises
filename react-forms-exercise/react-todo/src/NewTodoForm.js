import React, { useState } from "react";

const NewTodoForm = ({ addTodo }) => {
  const INITIAL_STATE = {
    todo: "",
  };

  const [formData, setFormData] = useState(INITIAL_STATE);
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    addTodo({ ...formData });
    setFormData(INITIAL_STATE);
  };
  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="todo">New ToDo</label>
      <input
        id="todo"
        type="text"
        name="todo"
        placeholder="Add ToDo"
        value={formData.todo}
        onChange={handleChange}
      />
      <button>Add Box</button>
    </form>
  );
};

export default NewTodoForm;
