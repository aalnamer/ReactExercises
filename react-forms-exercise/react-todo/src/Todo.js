import React from "react";

const Todo = ({ todo, id, handleRemove }) => {
  const remove = () => handleRemove(id);
  return (
    <ul className="todo">
      {todo}
      <div>
        <button onClick={remove}>Completed</button>
      </div>
    </ul>
  );
};

export default Todo;
