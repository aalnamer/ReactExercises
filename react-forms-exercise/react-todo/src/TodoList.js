import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Todo from "./Todo";

import NewTodoForm from "./NewTodoForm";

const TodoList = () => {
  const INITIAL_STATE = [];

  const [todos, setTodos] = useState(INITIAL_STATE);

  const addTodo = (newTodo) => {
    setTodos((todos) => [...todos, { ...newTodo, id: uuid() }]);
  };
  const remove = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };

  return (
    <div>
      <NewTodoForm addTodo={addTodo} />
      <div>
        {todos.map(({ id, todo }) => (
          <Todo id={id} todo={todo} handleRemove={remove} key={id} />
        ))}
      </div>
    </div>
  );
};

export default TodoList;
