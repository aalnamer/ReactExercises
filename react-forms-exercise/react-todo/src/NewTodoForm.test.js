import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";
import TodoList from "./TodoList";
import Todo from "./Todo";

//Smoke
it("renders without crashing", function () {
  render(<TodoList />);
});

//Snapshot
it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

// Use Form

it("should add a new  todo", function () {
  const { queryByText, getByLabelText } = render(<TodoList />);
  const input = getByLabelText("New ToDo");
  const btn = queryByText("Add ToDo");
  expect(queryByText("Completed")).not.toBeInTheDocument();
  fireEvent.change(input, { target: { value: "Clean" } });
  fireEvent.click(btn);
  expect(queryByText("Completed")).toBeInTheDocument();
});
