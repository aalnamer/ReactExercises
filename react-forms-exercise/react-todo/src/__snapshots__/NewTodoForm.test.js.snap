// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="todo"
      >
        New ToDo
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Add ToDo"
        type="text"
        value=""
      />
      <button>
        Add ToDo
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
